ENH: ExtensionArray.setitem

TomAugspurger opened the pull request:
xref https://github.com/pandas-dev/pandas/issues/19696

Adds `ExtensionBlock.__setitem__`.

This is only tested for with `DecimalArray` and `Categorical`. Supporting something like `JSONArray` where the "scalar" elements are actual sequences runs into issues elsewhere in internals. Maybe someday we can support that.

TomAugspurger committed:
ENH: ExtensionArray.setitem

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -874,22 +875,7 @@ def setitem(self, indexer, value, mgr=None):
         values = transf(values)
 
         # length checking
-        # boolean with truth values == len of the value is ok too
```
This was refactored out to `pandas/util/_validators` since I needed it in `ExtensionArray.setitem`.

TomAugspurger commented on pandas/core/internals.py:
```diff
@@ -3489,7 +3484,8 @@ def apply(self, f, axes=None, filter=None, do_integrity_check=False,
                             for k in align_keys
-                            if hasattr(kwargs[k], 'values'))
+                            if hasattr(kwargs[k], 'values') and
+                            not isinstance(kwargs[k], ABCExtensionArray))
```
If an ExtensionArray chooses to store it's data as `.values`, `setitem` would be broken without this extra check.

jorisvandenbossche replied:
Should we make a test for this? (eg call the underlying data `.values` in of the example test arrays?

jorisvandenbossche replied:
`_values`

jreback replied:
this is pretty special casey here. shouldn't this check for ``._values``?

TomAugspurger replied:
I'm not sure. I don't really know could be in kwargs. You think it's only ever Index or Series? Or could it be a dataframe or block?

TomAugspurger committed:
Avoid tm.assert

jreback commented on pandas/core/internals.py:
```diff
@@ -1898,6 +1884,15 @@ def is_view(self):
         return False
 
+    def setitem(self, indexer, value, mgr=None):
+        if isinstance(indexer, tuple):
```
can you add a doc-string

jreback commented on pandas/util/_validators.py:
```diff
@@ -356,3 +358,48 @@ def validate_fillna_kwargs(value, method, validate_scalar_dict_value=True):
+    """
+    from pandas.core.indexing import length_of_indexer
+
+    # boolean with truth values == len of the value is ok too
```
this is applicable in .where (in internals, but maybe also in frame.py). you *may* want to rename / clean that up.

TomAugspurger replied:
I didn't see anywhere in `Block.where` where similar logic was used.

jreback replied:
hmm maybe I cleaned this up a while back. IN any event this shouldn't be here at all. _validators is not the correct place. this is purely indexing validation. goes in core/indexing.py

jorisvandenbossche replied:
I can't really say something about where it belongs (would have to look into more detail), but note that it was copied straight from `internals.py`, and it is still called there. 
So in that sense this PR is just keeping the situation as it was.

TomAugspurger committed:
Docstring

jorisvandenbossche commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -0,0 +1,116 @@
+    def test_setitem_expand_columns(self, data):
+        df = pd.DataFrame({"A": data})
+        df['B'] = 1
+        assert len(df.columns) == 2
```
Can you add a test here for expanding with `data` ?

```
df = pd.DataFrame({"A": [1]*len(data)})
df['B'] = data
```

and also one that overwrites an existing column with `data` 

And for both, we should test both with `df['col']` as with `df.loc[:, 'col']`

TomAugspurger committed:
Alias for common data attributes

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+set

TomAugspurger committed:
Additional tests

TomAugspurger committed:
Linting

jreback commented on pandas/core/internals.py:
```diff
@@ -815,11 +816,24 @@ def _replace_single(self, *args, **kwargs):
+            The subset of self.values to set
+        value : object
+            The value being set
+        mgr : BlockPlacement
```
optional

jreback commented on pandas/core/internals.py:
```diff
@@ -1898,6 +1897,37 @@ def is_view(self):
+            The subset of self.values to set
+        value : object
+            The value being set
+        mgr : BlockPlacement
```
optional

jreback commented on pandas/core/internals.py:
```diff
@@ -1898,6 +1897,37 @@ def is_view(self):
+        'indexer' is a direct slice/positional indexer. 'value' must
+        be a compatible shape.
+        """
+        if isinstance(indexer, tuple):
```
hmm, you are doing to for compatibility? this is not very robust and should be handled by the caller. indexing is quite tricky as the contracts for who converts what are almost all handled in core/indexing.py So the internals are so of you-know-what-you-are doing. Since you are exposing this almost directly, I think you need some handling routines (and am talking about the indexers here) IN EA, which actually call things in core/indexing.py.

TomAugspurger replied:
> hmm, you are doing to for compatibility?

I guess, in the sense that things didn't work without it :) I think at some point, probably in `indexing.py`, the indexer is translated to a 2-D indexer, even though we know that EAs are always 1-D.

> Since you are exposing this almost directly

I don't think the call stack is any different for setting on an ExtensionBlock vs. anything else.

> I think you need some handling routines (and am talking about the indexers here) IN EA

IIUC, you're saying that the translation from a 2-D indexer to a 1-D indexer should happen elsewhere, yes? And the validation too? That sounds reasonable.

TomAugspurger replied:
@jreback what place in `core/indexing.py` did you have in mind? `_setitem_with_indexer`?

I don't see anywhere there that deals with specific blocks, just whether or not the block managner is holding mixed types (split path or not).

jorisvandenbossche commented on pandas/tests/extension/json/test_json.py:
```diff
@@ -71,3 +71,6 @@ def test_value_counts(self, all_data, dropna):
     pass
+
+# We intentionally don't run base.BaseSetitemTests because pandas'
+# internals has trouble setting sequences of values into scalar positions.
```
you can also add the test class as above, but skip the full class (I think the pytest skip decorators also work for classes)

TomAugspurger replied:
I've been scared off mixing inheritance and skipping classes since seeing pytest-568, where skipping in a child marks all the other children as skips.

jorisvandenbossche replied:
that's a good reason :-)

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+set

TomAugspurger commented:
This currently fails when the repr is triggered, not in the setitem as it should

```python
df = pd.DataFrame({"B": [1, 2, 3]})
df['A'] = IPArray([1, 2])
repr(df)
```

TomAugspurger commented:
What would we expect the type of this to be

```python
df = pd.DataFrame({"A": IPArray([1, 2, 3])})
df.loc[:, 'A'] = 1
```

This results in an integer dtype column. I think that's correct, right? The previous block is being entirely replaced? Then we have symmetry with `df['A'] = 1`.

TomAugspurger committed:
Move file

TomAugspurger committed:
BUG: Fixed length validation

TomAugspurger committed:
Removed print

TomAugspurger commented:
One unhandled case

```python
In [3]: s = pd.Series(IPArray([1, 2]), index=[(0, 1), (1, 2)])

In [4]: s[(0, 1)] = 2
```

```pytb
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/sandbox/pandas-ip/pandas/pandas/core/series.py in setitem(key, value)
    871             try:
--> 872                 self._set_with_engine(key, value)
    873                 return

~/sandbox/pandas-ip/pandas/pandas/core/series.py in _set_with_engine(self, key, value)
    930         try:
--> 931             self.index._engine.set_value(values, key, value)
    932             return

TypeError: Argument 'arr' has incorrect type (expected numpy.ndarray, got IPArray)

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-6c3a0d7e7408> in <module>()
----> 1 s[(0, 1)] = 2

~/sandbox/pandas-ip/pandas/pandas/core/series.py in __setitem__(self, key, value)
    922         # do the setitem
    923         cacher_needs_updating = self._check_is_chained_assignment_possible()
--> 924         setitem(key, value)
    925         if cacher_needs_updating:
    926             self._maybe_update_cacher()

~/sandbox/pandas-ip/pandas/pandas/core/series.py in setitem(key, value)
    904                 if (isinstance(key, tuple) and
    905                         not isinstance(self.index, MultiIndex)):
--> 906                     raise ValueError("Can only tuple-index with a MultiIndex")
    907
    908                 # python 3 type errors should be raised

ValueError: Can only tuple-index with a MultiIndex
```

The first exception is the problem. That cython method is expecting an ndarray, but we're giving it an ndarray or extensionarray.

TomAugspurger committed:
Fixed setitem for tuples

All our extension array setitem tests were hitting this. We only caught it
because only tuple reraised.

TomAugspurger commented:
Running asv with  43dfd7d   now.

TomAugspurger commented:
The last commit allowing EAs in the `index.engine.set*` had trouble with us using DatetimeIndex as a storage container for Datetime with TZ. We tried to directly modify the DatetimeIndex with `values[key] = value`. Previously this wasn't hit because we raised a TypeError earlier on, sending us down a different code path. I think that someday we'll want something like 43dfd7d, but not right now.

TomAugspurger committed:
Revert setitem changes

TomAugspurger commented:
Hmm, perhaps it's worth just punting on `__setitem__` on EAs with tuples for now. It's already failing on master for non-ndarrays

```python
In [1]: import pandas as pd

In [2]: arr = pd.date_range('2017', periods=4, tz='US/Eastern')

In [3]: s = pd.Series(arr, index=[(0, 1), (0, 2), (0, 3), (0, 4)])

In [4]: s[(0, 1)] = float('NaN')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/sandbox/pandas-ip/pandas/pandas/core/series.py in setitem(key, value)
    871             try:
--> 872                 self._set_with_engine(key, value)
    873                 return

~/sandbox/pandas-ip/pandas/pandas/core/series.py in _set_with_engine(self, key, value)
    930         try:
--> 931             self.index._engine.set_value(values, key, value)
    932             return

TypeError: Argument 'arr' has incorrect type (expected numpy.ndarray, got DatetimeIndex)

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-0a0fdcf8dcd6> in <module>()
----> 1 s[(0, 1)] = float('NaN')

~/sandbox/pandas-ip/pandas/pandas/core/series.py in __setitem__(self, key, value)
    922         # do the setitem
    923         cacher_needs_updating = self._check_is_chained_assignment_possible()
--> 924         setitem(key, value)
    925         if cacher_needs_updating:
    926             self._maybe_update_cacher()

~/sandbox/pandas-ip/pandas/pandas/core/series.py in setitem(key, value)
    904                 if (isinstance(key, tuple) and
    905                         not isinstance(self.index, MultiIndex)):
--> 906                     raise ValueError("Can only tuple-index with a MultiIndex")
    907
    908                 # python 3 type errors should be raised

ValueError: Can only tuple-index with a MultiIndex
```

With this hack, I can get it working

```python
diff --git a/pandas/core/series.py b/pandas/core/series.py
index e48012420..cd412844e 100644
--- a/pandas/core/series.py
+++ b/pandas/core/series.py
@@ -869,6 +869,18 @@ class Series(base.IndexOpsMixin, generic.NDFrame):
 
         def setitem(key, value):
             try:
+                if self._data.blocks[0].is_extension:
+                    raise TypeError
                 self._set_with_engine(key, value)
                 return
             except com.SettingWithCopyError:
```

I don't think that's worthwhile including to support tuples in the index. I'd prefer to see a proper fix for whatever is going wrong there.

TomAugspurger committed:
Xfail that test

TomAugspurger committed:
Linting

jorisvandenbossche commented:
> perhaps it's worth just punting on __setitem__ on EAs with tuples for now. It's already failing on master for non-ndarrays

No problem with that for now.

> [on df.loc[:, 'A'] = 1] This results in an integer dtype column. I think that's correct, right? The previous block is being entirely replaced? Then we have symmetry with df['A'] = 1.

I thought we somehow handled `df.loc[:, 'A'] = 1` and `df['A'] = 1` differently right now, but it seems that also in case of `df.loc[:, 'A'] = 1`  with 'normal' dtypes like int and float, the full block gets replaced. In that case, yes, I think we want to do the same here.

jorisvandenbossche commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -0,0 +1,146 @@
+        assert df.loc[10, 'B'] == data[1]
+
+    @pytest.mark.parametrize('as_callable', [True, False])
+    def test_set_mask_aligned(self, data, as_callable):
```
maybe call this `test_setitem_..` instead of `test_set_...` to be consistent with above

jorisvandenbossche replied:
or parametrize this for setitem and loc?

jorisvandenbossche commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -0,0 +1,146 @@
+        assert ser[0] == data[5]
+        assert ser[1] == data[6]
+
+    def test_set_mask_broadcast(self, data):
```
same here (set -> setitem)

jorisvandenbossche commented on pandas/tests/extension/base/setitem.py:
```diff
@@ -0,0 +1,146 @@
+
+        result = df.copy()
+        result.loc[:, 'B'] = 1
+        self.assert_frame_equal(result, expected)
```
maybe add overwriting the existing int `B` column with `data` ?

jorisvandenbossche reviewed (commented):
Looks good to me! 
Added some minor comments on the tests.

TomAugspurger committed:
Test updates

jreback commented on pandas/core/frame.py:
```diff
@@ -3204,7 +3204,10 @@ def reindexer(value):
         elif isinstance(value, ExtensionArray):
+            from pandas.core.series import _sanitize_index
             value = value.copy()
+            # Copy donesn't have any effect at the moment
```
why are you adding this comment? this is on-purpose

TomAugspurger replied:
The comment isn't clear, sorry. It's to answer "why call `value.copy()`" and then pass it to a function that takes a `copy` parameter. That's because `__sanitize_index` doesn't copy an EA, even with `copy=True`.

jreback commented on pandas/core/indexing.py:
```diff
@@ -2310,6 +2310,51 @@ def check_bool_indexer(ax, key):
+        When the indexer is an ndarray or list and the lengths don't
+        match.
+    """
+    from pandas.core.indexing import length_of_indexer
```
unecessary import

jreback commented on pandas/core/indexing.py:
```diff
@@ -2310,6 +2310,51 @@ def check_bool_indexer(ax, key):
+        match.
+    """
+    from pandas.core.indexing import length_of_indexer
+
```
you need to assert ``com.is_bool_indexer`` here (and you can actually remove some of this logic as it duplicates).

further should actually move ``is_bool_indexer`` to ``pandas.core.indexing``

jorisvandenbossche replied:
I think `com.is_bool_indexer` serves a different purpose, and should already have been called typically when ending up in this place. 
This helper function is only for generating informative error messages when the length does not match, `com.is_bool_indexer` does actual conversion and inference of the indexer.

jreback replied:
of course, but its serving a very similar purpose to this and needs to be moved here.

TomAugspurger replied:
`is_bool_indexer` seems to accept list-like, while this explicitly needs an ndarray since it checks `indexer[indexer]`. Given that I need to do `isinstance(indexer, ndarray)` anyway, to ensure that that boolean masking works, to tradeoff is

```python
            if not (isinstance(indexer, np.ndarray) and
                    indexer.dtype == np.bool_ and
                    len(indexer[indexer]) == len(value)):
 
```

vs.

```python
            if not (isinstance(indexer, np.ndarray) and
                    com.is_bool_indexer(indexer) and
                    len(indexer[indexer]) == len(value)):
 ```

Since the first is likely to be faster, I'd prefer to just go with that.

TomAugspurger committed:
Import, comment

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+set

TomAugspurger committed:
Removed the _values alias

that was causing issues with factorize.

TomAugspurger commented:
I had to remove the `_values` alias for now, as factorize uses the `getattr(values, '_values', values)` pattern to unbox a Series / Index.

jreback commented:
can you rebase

TomAugspurger commented:
I think all the tests have passed an there's no merge conflict.

On Sun, Apr 1, 2018 at 9:26 AM, Jeff Reback <notifications@github.com>
wrote:

> can you rebase
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/19907#issuecomment-377790457>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIrvrkqmLuCjxyufLE-LJRyKtIlhVks5tkOOMgaJpZM4STfE9>
> .
>

jreback commented:
@TomAugspurger 

- let's rebase one more time.
- doc-strings looks good, but pls give a once over
- I would collect all the issues that are EA and add onto the sub-section in whatsnew (just list them)

merge when ready-

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into fu1+set

TomAugspurger committed:
DOC: Fixup docstrings

jreback commented:
thanks @TomAugspurger
